nursery_store.go

```diff
@@ -0,0 +1,1377 @@
+//   the 'utxn' prefix. This allows multiple utxo nurseries for distinct chains
+//   to simultaneously use the same channel.DB instance. This is critical for
+//   providing replay protection and more to isolate chain-specific data in the
+//   multichain setting.
+//
+//   utxn<chain-hash>/
+//   |
+//   |   LAST FINALIZED HEIGHT
+//   |
+//   |   Each nursery store tracks a "last finalized height", which records the
+//   |   most recent block height for which the nursery store has purged all
+//   |   state. This value lags behind the best block height for reorg safety,
+//   |   and serves as a starting height for rescans after a restart.
+//   |
+//   ├── last-finalized-height-key: <last-finalized-height>
+//   |
+//   |   CHANNEL INDEX
+//   |
+//   |   The channel index contains a directory for each channel that has a
+//   |   non-zero number of outputs being tracked by the nursery store.
+//   |   Inside each channel directory are files contains serialized spendable
+//   |   outputs that are awaiting some state transition. The name of each file
+//   |   contains the outpoint of the spendable output in the file, and is
+//   |   prefixed with 4-byte state prefix, indicating whether the spendable
+//   |   output is a crib, preschool, or kindergarten output. The nursery store
+//   |   supports the ability to enumerate all outputs for a particular channel,
+//   |   which is useful in constructing nursery reports.
+//   |
+//   ├── channel-index-key/
+//   │   ├── <chain-point-1>/                      <- CHANNEL BUCKET
+//   |   |   ├── <state-prefix><outpoint-1>: <spendable-output-1>
+//   |   |   └── <state-prefix><outpoint-2>: <spendable-output-2>
+//   │   ├── <chain-point-2>/
+//   |   |   └── <state-prefix><outpoint-3>: <spendable-output-3>
+//   │   └── <chain-point-3>/
+//   |       ├── <state-prefix><outpoint-4>: <spendable-output-4>
+//   |       └── <state-prefix><outpoint-5>: <spendable-output-5>
+//   |
+//   |   HEIGHT INDEX
+//   |
+//   |   The height index contains a directory for each height at which the
+//   |   nursery still has uncompleted actions. If an output is a crib or
+//   |   kindergarten output, it will have an associated entry in the height
+//   |   index. Inside a particular height directory, the structure is similar
+//   |   to that of the channel index, containing multiple channel directories,
+//   |   each of which contains subdirectories named with a prefixed outpoint
+//   | 	 belonging to the channel. Enumerating these combinations yields a
+//   |   relative file path:
+//   |     e.g. <chan-point-3>/<prefix><outpoint-2>/
+//   |   that can be queried in the channel index to retrieve the serialized
+//   |   output.
+//   |
+//   └── height-index-key/
+//       ├── <height-1>/                             <- HEIGHT BUCKET
+//       |   └── <chan-point-3>/                     <- HEIGHT-CHANNEL BUCKET
+//       |   |    ├── <state-prefix><outpoint-4>/    <- PREFIXED OUTPOINT
```

Comment:
Why are these buckets instead of keys with empty values? According to Bolt docs: "It's important to note that you can have a zero-length value set to a key which is different than the key not existing."

Actually, taken a step further, I think you can remove two layers of bucketing with a structure like:

```
height-index-key/
|---- <height-1>/
|     |---- <state-prefix><outpoint>: <chan-point>
```

Reply:
The decision between buckets and files here was sort of arbitrary. When iterating over the buckets, you get `nil` values, instead of zero-length byte slices, so not a huge difference there. The primary reason I chose buckets is it could allow us to add metadata files in the height index w/o introducing breaking changes, but don't really have a strong preference either way.

I still think we should stick to the two-layer bucketing approach, as the proposed single-layer structure will require more on-disk storage, and tends to go against the advice of traditional database normalization strategies concerning how to store functional dependencies.

Consider a nursery store with `n` channels and `m` outputs each, so `n*m` outputs in total; each prefixed key occupies `op = 4+66 bytes`, and each chan point takes `cp = 66 bytes`. 

For now, assume all outputs need to be spent at one height. Using the two-layer design, we get `T2 = n*cp + n*m*op = n * (cp + m*op)`. This represents the best case space storage for the two-layer design, since each chan point and outpoint is serialized exactly once.

With the single layer design, this total size is `T1 = n*m*(cp + op)`, since an prefixed key and channel point are serialized for every outpoint. Taking the difference, we get `T2 - T1 = n * (m - 1) * cp`. For something like `n = 10` and `m = 200`, this gives `T1 = 272K`, `T2 = 141K`, and `T2-T1 = 131K`, roughly doubling the size of the best-case, two-layered height index.

Note that the lower bound for `T1` is _independent of how the outputs are distributed across the height buckets_.  The two-layer design only degrades to this in the case that a channel has _at most_ one output at a particular height. Due to the nature of channel closes, I think we will see many tight groupings of outputs being entered into the height index, so I suspect, in practice, that the two-layered approach will tend closer to the best case.

Reply:
OK, that totally makes sense in terms of keeping it nested. However, I would kind of prefer for the leaves to be empty values instead of buckets. Looking briefly at [boltdb](https://github.com/boltdb/bolt/blob/master/bucket.go), there's definitely some overhead in storing buckets. Also, depending on how much metadata you are imagining, if it can be serialized as a byte vector, it could always just be stored as the value in the future without breaking compatibility. Up to you though of course, doesn't make a huge difference, as you said.

Reply:
Let's do the files with empty values, it will better mirror the channel bucket structure, possibly allowing some more code reuse between the two!